### ISSUE TITLE
having trouble verifying #1119 fixes (or doesn't) #1115 (or #1113)

### ISSUE THREAD

**Issue by ConceptualWempa** [mamba-org/mamba#1133]
Sorry for submitting a new issue for this, but it wasn't obvious how else to do it since the bug reports and fix are all closed.

I am trying to verify that change #1119 fixes the case described in #1115 ... on the surface it doesn't appear to, but I'm not certain I've successfully built and installed the revision in question, and e.g. 'mamba info' doesn't show anything that tips off that I've got a custom build and not "vanilla" 0.15.2 - is there a good way to tell, short of e.g. lsof?  mamba is also present in the base environment, but that version is 0.11.1, so it would be (or should be?) obvious if I was mistakenly invoking that.

steps I did:
- git clone https://github/mamba-org/mamba.git
- git checkout e5f007bf
- mamba env create --name mamba-dev --file environment-dev.yml
- conda activate mamba-dev
- conda install conda
- pip install -e .

> echo $CONDA_PREFIX
/hostname/forte/conda-envs/envs/mamba-dev
> type mamba
mamba is /hostname/forte/conda-envs/envs/mamba-dev/bin/mamba
> mamba --version
mamba 0.15.2
conda 4.10.3
> mamba --no-banner create -n test python==2.7

Looking for: ['python==2.7']

pkgs/main/linux-64       [====================] (00m:26s) 0 Failed
...
    RuntimeError: Download error (28) Timeout was reached [https://repo.anaconda.com/pkgs/main/noarch/repodata.json]

(left out most of the error report, for brevity), whereas 'conda info' shows:

           channel URLs : https://artifactory/artifactory/api/conda/remote-conda/main/linux-64
                          https://artifactory/artifactory/api/conda/remote-conda/main/noarch

but 'mamba info' shows:

           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch

(which is the same behavior as 0.15.2 from conda-forge ... so no clue there).

conda config:
> conda config --show-sources
==> /opt/ci1/toolchain/conda/4.10.0/condarc <==
auto_update_conda: False
notify_outdated_conda: False
aggressive_update_packages: []
envs_dirs:
  - /hostname/${USER}/conda-envs/envs
pkgs_dirs:
  - /hostname/${USER}/conda-envs/pkgs
proxy_servers:
  http:
  https:
channel_alias: https://artifactory/artifactory/api/conda/remote-anaconda-cloud-conda/
channels:
  - defaults
custom_channels:
  in_house: https://artifactory/artifactory/api/conda/local-conda
  3rd_party: https://artifactory/artifactory/api/conda/local-conda
  experimental: https://artifactory/artifactory/api/conda/local-conda
default_channels:
  - https://artifactory/artifactory/api/conda/remote-conda/main
show_channel_urls: True
report_errors: False
anaconda_upload: False

**Comment by wolfv** [comment]
I've just released mamba. Can you try with the latest release? (0.15.3)

**Comment by ConceptualWempa** [comment]
sorry for the delay.

Yes, 0.15.3 does seem to fix the behavior, in that, for an environment with default_channels overridden in that environment's condarc:

mamba update --all

- for mamba==0.15.2, times out trying to reach anaconda.org.
- for mamba==0.15.3, resolves and presents a list of packages to update from our artifactory "mirror".

I think I got snagged (when attempting to test pre-release) by a different behavior difference between conda and mamba, which may or may not be a bug ... conda (as of 4.10.3) always takes the base environment's condarc into account, even if a sub-environment that has its own copy of conda installed is active.  For the equivalent situation (sub-env has mamba installed ... which also implies conda installed), mamba considers the sub-env to be 'base', and ignores the "real" base environment's condarc.  This does not appear to be new behavior (confirmed at least as far back as 0.11.1), and I suspect it is intentional ... it does appear to contradict the description in the mamba docs, though:

https://mamba.readthedocs.io/en/latest/user_guide/configuration.html#multiple-rc-files

which may just mean that bit of the documentation is erroneous/stale.

mamba shows the shell level identically to conda, at least for the cases I have tried, i.e.

- 0 == no env active ('conda' command available due to sourcing etc/profile.d/conda.sh from base env; see (*) below for 'mamba')
- 1 == an env has been activated; no other activation events have occurred
- 2 == another env was activated after the initial activation

assuming level 1 is the base env (i.e. 'conda activate'), and level 2 is some sub-env of that base with mamba installed, conda and mamba behave identically (for condarc resolution purposes) at both level 0* and level 1, but mamba ignores the base env's condarc at level 2 (conda, of course, does not).  Mamba reports the active environment as "base" at level 2, which is consistent with it ignoring the "real" base env's condarc ... but in that case, mamba should probably show the shell level as 1, since it's effectively as if the first activation never happened.

* mamba showing level 0 is a bit of a 'cheat' case - it's not in the "path" (either literally or as a shell function) before any envs have been activated; it's still in the (literal) path after activating at least once and then deactivating all environments, due to conda's special treatment of 'condabin'.   The upshot of which is, mamba can be made to show level 0 (no envs active) with the "cheat" of having activated/deactivated at least once (or by simply calling it via the full path) ... for the purposes of this illustration, I've assumed that the same base env was "pre-activated" to get that result, otherwise who knows what might happen.

**Comment by wolfv** [comment]
@ConceptualWempa did you install mamba in a non-base env? You need to install mamba side-by-side with conda in the conda base env, otherwise behaviors can occur like what you're describing.

**Comment by ConceptualWempa** [comment]
right, exactly; when intially probing the problem described in #1115, I did have mamba installed in the base env.  when attempting to validate #1119's fix I did not, because I was working in a dev environment.

so the solution to that problem is either, as you say, "keep" mamba in the base environment, or else the base environment's condarc contents have to "follow" mamba (i.e. I'd have to copy it to whichever environment I wanted to test and possibly modify mamba in, possibly with further modifications, because at that point I can't rely on my personal .condarc ... at least, not if that diagram in the docs is correct.  Though I was under the impression that, at least for conda, the precedence order is [lower] base env -> target env -> user profile [higher] ... I'm far from certain of that, though).

It was just very confusing until I dug in deeper, because both situations produced very similar results, for very different reasons.  ;-)

that said, my only lingering question here is:

is the behavior of mamba (to ignore the base env's condarc once a non-base env is activated) intentional?  (in which case the doc is a bit misleading) Or is the doc correct and the behavior an "improvement opportunity"?  If the latter, I might take a stab at fixing, unless it's a quick and obvious change for you.
